pyogrio/_io.pyx

```diff
@@ -1678,6 +1678,117 @@ cdef infer_field_types(list dtypes):
 
     return field_types
 
+def ogr_write_arrow(
+    obj,
+    str path
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    stream_capsule = obj.__arrow_c_stream__()
+    return write_arrow_stream_capsule(stream_capsule)
+
+cdef OGRErr write_arrow_stream_capsule(OGRLayerH destLayer, object capsule):
+    cdef ArrowArrayStream stream
+    cdef ArrowSchema schema
+    cdef ArrowArray array
+
+    stream = PyCapsule_GetPointer(capsule, "arrow_array_stream")
+    if not stream:
+        raise RuntimeError("Accessing PyCapsule pointer named 'arrow_array_stream' failed.")
+
+    if stream.release == NULL:
+        raise RuntimeError("Arrow Array Stream was already released.")
+
+    errcode = stream.get_schema(stream, &schema)
+    if errcode != 0:
+        stream.release(stream)
+        raise RuntimeError("Error while accessing schema from stream.")
+
+    errcode = create_fields_from_arrow_schema(destLayer, &schema)
+    if errcode != 0:
+        schema.release(&schema)
+        stream.release(stream)
+        raise RuntimeError("Error creating Arrow Schema in OGR layer.")
+
+    while True:
+        errcode = stream.get_next(stream, &array)
+        if errcode != 1:
+            schema.release(&schema)
+            stream.release(stream)
+            raise RuntimeError("Error while accessing batch from stream.")
+
+        # We've reached the end of the stream
+        if array.release == NULL:
+            break
+
+        errcode = OGR_L_WriteArrowBatch(destLayer, &schema, &array, options)
+        if errcode:
+            if array.release != NULL:
+                array.release(&array)
+
+            schema.release(&schema)
+            stream.release(stream)
+            raise RuntimeError("Error while writing batch to OGR layer.")
+
+        if array.release != NULL:
+            array.release(&array)
+
+
+    schema.release(&schema)
+    stream.release(stream)
+
+# Create output fields using CreateFieldFromArrowSchema()
+static bool create_fields_from_arrow_schema(
+    OGRLayerH destLayer,
+    const struct ArrowSchema* schema,
+    char** options
+):
+    # The schema object is a struct type where each child is a column.
+    for child in schema.n_children:
+        # Access the metadata for this column
+        const char *metadata = child.metadata
+
+        # TODO: I don't know how to parse this metadata in C... I guess I can just use Python APIs for this in Cython?
+        # https://github.com/OSGeo/gdal/pull/9133/files#diff-37bedc92ae1d5e04706c7b9f8ea9e9fcccf984ca0c9997e2020ff85f1b958433R1159-R1185
+        # if metadata:
```

Comment:
Hmm, it's a pity that GDAL doesn't provide a helper to create a full layer definition from a schema, instead of only field by field (that would us to access the individual children of the ArrowSchema). 

We might want to vendor some helpers from nanoarrow-c to extract the children, and the metadata etc.

Reply:
Accessing the child is something we can probably do quite easily, as it's something like `schema.children[i]`

For a first prototype, I think we could also ignore the metadata for a moment, and manually specify which are the geometry columns.